src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs

```diff
@@ -100,7 +101,8 @@ public static class LazyInitializer
         /// </para>
         /// </remarks>
         public static T EnsureInitialized<T>([NotNull] ref T? target, Func<T> valueFactory) where T : class =>
-            Volatile.Read(ref target!) ?? EnsureInitializedCore(ref target, valueFactory);
+            target ??
+            EnsureInitializedCore(ref target, valueFactory);
```

Comment:
I see what you meant in your question about volatile reads inside the Lazy init - what if the initializer does some random writes unrelated to the instance, can we see uninitialized data.

Basically if T's constructor initializes static s, otherwise not related to T.

```cs
if (instT != null) s.ToString();   // no guarantees, can read s before instT

if (EnsureInitialized(ref instT) != null) s.ToString();   // seems like s will be initialized
```

I do not think we document `EnsureInitialized` as an acquiring read, but implementation is an acquiring read. 
Relaxing that might be observable. 

Is this API old enough that someone might depend on this being an acquire?

Reply:
After thinking about it, perhaps the `Volatile.Read` here and `Volatile.Write` in EnsureInitializedCore are intentional - to include all sideeffects of constructor/factory under "initialized" state? 

Especially in the factory case - I can imagine the factory to do a lot of things and then just assign a trivial object as a marker that initialization is done.

Reply:
@stephentoub - just wonder what the final thinking on this. Are we ok with relaxing the guarantees to just the data reachable via the instance?

Reply:
Somehow it was showing to me that the change is still there. Now it looks reverted.